### PR TITLE
AAE

### PR DESCRIPTION
**Pull request by vitskvara**
Adversarial autoencoders.

### CODE REVIEW

File: src/models/aae.jl [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+end
+
+"""
+	AAE(prior, encoder, decoder, discriminator)
+	AAE(zdim::Int, encoder, decoder, discriminator)
+
+Constructs an adversarial autoencoder.
+"""
+function AAE(zdim::Int, encoder, decoder, discriminator)
+	W = first(Flux.params(encoder))
```
Comment (janfrancu) [inline src/models/aae.jl:15]: Align.

File: src/models/vae.jl [hunk 2]
```diff
@@ -14,6 +16,16 @@ Safe version of softplus.
...
 
+"""
+	init_vamp(pseudoinput_mean,k::Int)
+
+Initializes the VAMP prior from a mean vector and number of components.
+"""
+function init_vamp(pseudoinput_mean, k::Int)
+	T = eltype(pseudoinput_mean)
+	pseudoinputs = T(1) .* randn(T, size(pseudoinput_mean)[1:end-1]..., k) .+ pseudoinput_mean
+	VAMP(pseudoinputs)
```
Comment (janfrancu) [inline src/models/vae.jl:27]: Out of curiosity. Where is VAMP imported from?

File: src/models/aae.jl [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+#loss(m::AAE,x,batchsize::Int) = aeloss(m,x,batchsize) + dloss(m,x,batchsize) + gloss(m,x,batchsize)
+
+"""
+	StatsBase.fit!(model::AAE, data::Tuple; max_train_time=82800, 
+	lr=0.001, batchsize=64, patience=30, check_interval::Int=10, kwargs...)
+"""
+function StatsBase.fit!(model::AAE, data::Tuple; max_train_time=82800, lr=0.001, 
+	batchsize=64, patience=30, check_interval::Int=10, kwargs...)
+	history = MVHistory()
+	aeopt = ADAM(lr)
```
Comment (janfrancu) [inline src/models/aae.jl:148]: Why do you have more than one optimizer, if they just share the same learning rate?

File: src/models/aae.jl [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+			_patience = patience
+
+			# this should save the model at least once
+			# when the validation loss is decreasing 
+			if mod(i, 10) == 0
+				model = deepcopy(tr_model)
+			end
+		elseif time() - start_time > max_train_time # stop early if time is running out
+			model = deepcopy(tr_model)
+			@info "Stopped training after $(i) iterations, $((time() - start_time)/3600) hours."
```
Comment (janfrancu) [inline src/models/aae.jl:217]: Good idea to print the time it took.

File: src/models/aae.jl [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+	# again, this is not optimal, the model should be passed by reference and only the reference should be edited
+	(history=history, iterations=i, model=model, npars=sum(map(p->length(p), Flux.params(model))))
+end
+
+"""
+	AEModel
+
+A Union of VAE and AAE types.
+"""
+AEModel = Union{VAE, AAE}
```
Comment (janfrancu) [inline src/models/aae.jl:237]: This may be not the best place for the shared code, but let's leave it here.

**Review by janfrancu — CHANGES_REQUESTED** [review]

File: src/models/aae.jl (hunk 1 above)
Comment (vitskvara) [reply src/models/aae.jl:15]: Alright

File: src/models/vae.jl (hunk 2 above)
Comment (vitskvara) [reply src/models/vae.jl:27]: It's from GenerativeModels.jl

File: src/models/aae.jl (hunk 3 above)
Comment (vitskvara) [reply src/models/aae.jl:148]: Because each optimizer optimizes different parts of the model and different set of parameters, therefore the history (moments etc.) is different. Or can this be achieved by using a single optimizer for all parameters?

File: src/models/aae.jl (hunk 5 above)
Comment (vitskvara) [reply src/models/aae.jl:237]: Yeah it's not really a lot of code so I thought it does not really need a separate file.

File: src/models/aae.jl (hunk 3 above)
Comment (janfrancu) [reply src/models/aae.jl:148]: Don't know. 
If it works, don't change it. :smile:

**Review by janfrancu — APPROVED** [review]

File: src/models/aae.jl (hunk 3 above)
Comment (vitskvara) [reply src/models/aae.jl:148]: originally, I only had one optimizer and was surprised that it worked at all - maybe it is restarted everytime it encounters new parameters? But the convergence is much better with separate optimizers.
